library/core/src/ops/control_flow.rs

```diff
@@ -0,0 +1,67 @@
+use crate::ops::Try;
+
+/// Used to make try_fold closures more like normal loops
+#[unstable(feature="control_flow_enum", reason="new API", issue="75744")]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum ControlFlow<C, B> {
+    /// Continue in the loop, using the given value for the next iteration
+    Continue(C),
+    /// Exit the loop, yielding the given value
+    Break(B),
+}
+
+#[unstable(feature="control_flow_enum", reason="new API", issue="75744")]
+impl<C, B> Try for ControlFlow<C, B> {
+    type Ok = C;
+    type Error = B;
+    #[inline]
+    fn into_result(self) -> Result<Self::Ok, Self::Error> {
+        match self {
+            ControlFlow::Continue(y) => Ok(y),
+            ControlFlow::Break(x) => Err(x),
+        }
+    }
+    #[inline]
+    fn from_error(v: Self::Error) -> Self {
+        ControlFlow::Break(v)
+    }
+    #[inline]
+    fn from_ok(v: Self::Ok) -> Self {
+        ControlFlow::Continue(v)
+    }
+}
+
+impl<C, B> ControlFlow<C, B> {
+    /// Converts the `ControlFlow` into an `Option` which is `Some` if the
+    /// `ControlFlow` was `Break` and `None` otherwise.
+    #[inline]
+    #[unstable(feature="control_flow_enum", reason="new API", issue="75744")]
+    pub fn break_value(self) -> Option<B> {
+        match self {
+            ControlFlow::Continue(..) => None,
+            ControlFlow::Break(x) => Some(x),
+        }
+    }
+}
+
+impl<R: Try> ControlFlow<R::Ok, R> {
```

Comment:
Shouldn't the second parameter be `R::Err`?

Reply:
Yes, definitely.

Reply:
Oh, I see - we can't do this because it's an unconstrained type parameter. Is there a good way to solve that?

Reply:
No, the break part here is intentionally the full `impl Try` type (`Result`, etc).  They're definitely strange APIs, though -- `break_value` is something that could plausibly stabilize, but these ones probably aren't.

Maybe leave this particular impl block over in `iter` as non-pub so they can still be used there, but don't show up in the rustdoc?  (And won't need `#[unstable]` since they'll be unusable outside `iter`.)

Reply:
Oh, I see. I misunderstood how these were being used. You're right, it's a very strange API.